src/app/core/store/shopping/products/products.effects.ts

```diff
@@ -169,18 +172,44 @@ export class ProductsEffects {
     )
   );
 
-  loadProductBundles$ = createEffect(() =>
+  loadProductParts$ = createEffect(() =>
     this.actions$.pipe(
-      ofType(loadProductSuccess),
-      mapToPayloadProperty('product'),
-      filter(product => ProductHelper.isProductBundle(product)),
-      mergeMap(({ sku }) =>
-        this.productsService.getProductBundles(sku).pipe(
-          mergeMap(({ stubs, bundledProducts }) => [
-            ...stubs.map((product: Product) => loadProductSuccess({ product })),
-            loadProductBundlesSuccess({ sku, bundledProducts }),
-          ]),
-          mapErrorToAction(loadProductFail, { sku })
+      ofType(loadProductParts),
+      mapToPayloadProperty('sku'),
+      groupBy(identity),
+      mergeMap(group$ =>
+        group$.pipe(
+          this.throttleOnBrowser(),
+          mergeMap(sku =>
+            this.store.pipe(
+              select(getProductParts(sku)),
+              first(),
+              filter(parts => !parts?.length),
+              switchMap(() => this.store.pipe(select(getProduct(sku)))),
+              filter(product => ProductHelper.isProductBundle(product) || ProductHelper.isRetailSet(product)),
+              take(1)
+            )
+          ),
+          exhaustMap(product =>
+            ProductHelper.isProductBundle(product)
+              ? this.productsService.getProductBundles(product.sku).pipe(
+                  mergeMap(({ stubs, bundledProducts: parts }) => [
+                    ...stubs.map((stub: Product) => loadProductSuccess({ product: stub })),
```

Comment:
is it guaranteed that stubs will always contain values? If not, the [array spread operator might cause problems when used with an undefined value](https://github.com/Microsoft/TypeScript/issues/15027).

Reply:
It is defined because the mapper will at least yield an empty array. It would be a ICM configuration issue otherwise, if no parts are available. So an unexpected error that is not swallowed is also good. Don't you think? 😅